### PR TITLE
Fix thread pool executors to operate as intended

### PR DESCRIPTION
**Pull request by timw**
What does this PR do?

- Consolidates all construction of ThreadPoolExecutors into a single factory class that standardises the logging, thread naming and thread group management (which was inconsistent across many of the instances)
- Fixes many of the thread pool instances to scale as intended 

Motivation
Most of the thread pool executors in OrientDB look like they were designed to scale, but because of the peculiarities of the JDK ThreadPoolExecutor, do not.

Additional Notes

The JDK ThreadPoolExecutor has a very unintuitive API, where the behaviour of the BlockingQueue used by the executor determines whether the pool will scale beyond the core (up to the max threads). Only a queue that fails on `offer` will cause growth of the pool beyond the core, so any executor using an unbounded LinkedBlockingQueue (for example) would never grow beyond core.

I've factored all the instantiations into a single factory class, and created thread pool implementations that actually behave as the construction seems to have intended (except for specific cases as noted below). Someone more familiar with the parts of the OrientDB engine should review to ensure this was the intended behaviour, or whether a different behaviour is required.

The scaling implementation uses a custom blocking queue that tricks the ThreadPoolExecutor into scaling based on a target queue size. The target queue size can be as low as 0 (which will force the thread pool to grow in favour of queuing), and triggers the thread pool probabilistically as the queue grows closer to the target size.

I haven't tried to implement blocking - this is surprisingly hard to achieve safely given all the edge cases in the thread pool lifecycle.

Notes on specific thread pools:

- `OrientDBEmbedded` - intent seems to have been to scale from 1 to available processors, but would only ever allocate 1 thread. It now scales as intended, with a target queue size of 100. 
- `OWowCache` - this is simply a cached thread pool construct
- `Orient` - this made an attempt to implement a blocking construct, but because the override of `offer` never returned false, the pool size would never scale beyond the core size. It now scales as intended but will queue instead of blocking.
- `ODistributedDatabaseImpl` - this was intended to scale to `totalWorkers` but would never grow past 1 thread. It now scales between 1 and `totalWorkers`, with a target queue size of 0 forcing thread scaling in preference to queuing (since these are latency sensitive).
- `OPushManager` - this scaled from 0 to 5 threads, but would reject executions. New implementation behaves as previously.

Checklist
[x] I have run the build using `mvn clean package` command
[x] My unit tests cover both failure and success scenarios

### CODE REVIEW

File: .sdkmanrc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Enable auto-env through the sdkman_auto_env config
```
Comment (tglman) [inline .sdkmanrc]: Is this file needed for some reasons ?

File: core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    return newSingleThreadScheduledPool(threadName, Thread.currentThread().getThreadGroup());
+  }
+
+  public static ScheduledExecutorService newSingleThreadScheduledPool(
+      String threadName, ThreadGroup parentThreadGroup) {
+    return new OScheduledThreadPoolExecutorWithLogging(
+        1, new SingletonNamedThreadFactory(threadName, parentThreadGroup));
+  }
+
+  private abstract static class BaseThreadFactory implements ThreadFactory {
```
Comment (tglman) [inline core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java]: Same better in a specific file

File: core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+      final Thread thread = new Thread(parentThreadGroup, r);
+      thread.setDaemon(true);
+      thread.setName(nextThreadName());
+      return thread;
+    }
+
+    protected abstract String nextThreadName();
+  }
+
+  private static final class SingletonNamedThreadFactory extends BaseThreadFactory {
```
Comment (tglman) [inline core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java]: Same better in a specific file

File: core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+      this.name = name;
+    }
+
+    @Override
+    protected String nextThreadName() {
+      return name;
+    }
+  }
+
+  private static final class NamedThreadFactory extends BaseThreadFactory {
```
Comment (tglman) [inline core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java]: Same better in a specific file

File: core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+     * <br>
+     * The work queue will probabilistically reject {@link #offer(Runnable) offers} to the work
+     * queue (with that probability growing to 100% as the indicated capacity is reached). The
+     * rejection of the offer will trigger pool growth in the thread pool, and once the max pool
+     * size has been reached will cause {@link RejectedExecutionHandler#rejectedExecution(Runnable,
+     * ThreadPoolExecutor) rejection}. <br>
+     * The implementation of {@link RejectedExecutionHandler} for this queue will then add it
+     * directly to the work queue, bypassing the rejection logic.
+     */
+    private static class ScalingQueue extends LinkedBlockingQueue<Runnable> {
```
Comment (tglman) [inline core/src/main/java/com/orientechnologies/common/thread/OThreadPoolExecutors.java]: I would prefer this inner class would go to a specific file

File: core/src/main/java/com/orientechnologies/orient/core/Orient.java [hunk 6]
```diff
@@ -271,24 +271,8 @@ public Orient startup() {
       final int cores = Runtime.getRuntime().availableProcessors();
 
       workers =
-          new OThreadPoolExecutorWithLogging(
```
Comment (tglman) [inline core/src/main/java/com/orientechnologies/orient/core/Orient.java:274]: Will the new implementation be blocking ? looking at I think is not, would be better that this executor is kept blocking, may create some behaviour problem changing it to not blocking.

**Review by tglman — COMMENTED** [review]
Hi,

This looks good, just a small set of corrections, as well will run some additional tests on our side before merge this, anything I think this is solving some side problem in scheduling that I experienced in my tests, thank you for the work!!

File: .sdkmanrc (hunk 1 above)
Comment (timw) [reply .sdkmanrc]: No, that shouldn't have been committed.

File: core/src/main/java/com/orientechnologies/orient/core/Orient.java (hunk 6 above)
Comment (timw) [reply core/src/main/java/com/orientechnologies/orient/core/Orient.java:274]: That was one of the questions I had - I've added blocking behaviour back in now.

**Comment by timw** [comment]
Inner classes have been moved to top level, and blocking functionality reinstated.

The blocking is done as it was before - it's not a generally safe implementation (for that you need to handle a whole lot of task cancellation/shutdown/termination scenarios) but it's similar to what was there before, except now the thread pool will scale before growing the queue to the max queue size and then blocking.
